### PR TITLE
feat：Add recipient batch CSV batch import, dynamic import and export

### PR DESCRIPTION
**Pull request by wzh425**

### CODE REVIEW

File: src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/MessageTaskDto.cs [hunk 1]
```diff
@@ -21,6 +21,8 @@ public class MessageTaskDto : AuditEntityDto<Guid, Guid>
 
     public ReceiverTypes ReceiverType { get; set; }
 
+    public MessageTaskReceiverSelectTypes ReceiverSelectType { get; set; }
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Dtos/MessageTasks/MessageTaskDto.cs:24]: SelectReceiverTypes

File: src/Contracts/Masa.Mc.Contracts.Admin/Enums/MessageTasks/MessageTaskReceiverSelectTypes.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Mc.Contracts.Admin.Enums.MessageTasks;
+
+public enum MessageTaskReceiverSelectTypes
```
Comment (MayueCif) [inline src/Contracts/Masa.Mc.Contracts.Admin/Enums/MessageTasks/MessageTaskReceiverSelectTypes.cs:6]: SelectReceiverTypes

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    /// 导出字节
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="dataItems">数据列</param>
+    /// <returns></returns>
+    public Task<byte[]> ExportAsByteArray<T>(ICollection<T> dataItems) where T : class, new()
+    {
+        var helper = new ExportHelper<T>();
+        return Task.FromResult(helper.GetCsvExportAsByteArray(dataItems));
+
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs:44]: delete line

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvImporter.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    /// <param name="fileName"></param>
+    /// <returns></returns>
+    public async Task<ExportFileInfo> GenerateTemplate<T>(string fileName) where T : class, new()
+    {
+        using (var importer = new ImportHelper<T>(fileName))
+        {
+            return (await importer.GenerateTemplateByte())
+                .ToCsvExportFileInfo(fileName);
+        }
+    }
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvImporter.cs:21]: add blank line

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/Utility/ExportHelper.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    public ExportHelper()
+    {
+    }
+
+    /// <summary>
+    /// </summary>
+    /// <param name="type"></param>
+    public ExportHelper(Type type)
+    {
+        this._type = type;
```
Comment (MayueCif) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/Utility/ExportHelper.cs:25]: remove this.

File: src/Web/Masa.Mc.Web.Admin/Pages/MessageTasks/Modules/MessageReceiversSelect.razor.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                return false;
+            }
+            if (Value.Any(x => x.Email == receiver.Email))
+            {
+                await WarningAsync("邮箱不允许重复");
+                return false;
+            }
+            Value.Insert(0, receiver);
+        }
+        await ValueChanged.InvokeAsync(Value);
```
Comment (MayueCif) [inline src/Web/Masa.Mc.Web.Admin/Pages/MessageTasks/Modules/MessageReceiversSelect.razor.cs:94]: if(ValueChanged.HasDelegate()){
 await ValueChanged.InvokeAsync(Value);
}
else{
//todo
}

**Review by MayueCif — COMMENTED** [review]
modify

**Review by MayueCif — APPROVED** [review]

**Comment by MayueCif** [comment]
@doddgu LGTM

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    /// <param name="fileName">文件名</param>
+    /// <param name="dataItems">数据列</param>
+    /// <returns>文件</returns>
+    public async Task<ExportFileInfo> Export<T>(string fileName, ICollection<T> dataItems) where T : class, new()
+    {
+        fileName.CheckCsvFileName();
+        var bytes = await ExportAsByteArray(dataItems);
+        return bytes.ToCsvExportFileInfo(fileName);
+    }
+    /// <summary>
```
Comment (doddgu) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs:21]: add blank line

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    /// <param name="fileName"></param>
+    /// <param name="dataItems"></param>
+    /// <returns></returns>
+    public async Task<ExportFileInfo> Export<T>(string fileName, DataTable dataItems) where T : class, new()
+    {
+        fileName.CheckCsvFileName();
+        var bytes = await ExportAsByteArray<T>(dataItems);
+        return bytes.ToCsvExportFileInfo(fileName);
+    }
+    /// <summary>
```
Comment (doddgu) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs:34]: add blank line

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    /// <param name="dataItems"></param>
+    /// <returns></returns>
+    public Task<byte[]> ExportAsByteArray<T>(DataTable dataItems) where T : class, new()
+    {
+        var helper = new ExportHelper<T>();
+        return Task.FromResult(helper.GetCsvExportAsByteArray(dataItems, null));
+    }
+
+    /// <summary>
+    ///
```
Comment (doddgu) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs:58]: remove the comment

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
+﻿namespace Masa.Mc.Infrastructure.ExporterAndImporter.Csv;
+
+/// <summary>
+/// Csv导出程序
```
Comment (doddgu) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvExporter.cs:4]: English comment

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvImporter.cs [hunk 11]
```diff
@@ -0,0 +1,91 @@
+﻿namespace Masa.Mc.Infrastructure.ExporterAndImporter.Csv;
+
+/// <summary>
+///     Csv导入
```
Comment (doddgu) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/CsvImporter.cs:4]: English comment or remove it.

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/Masa.Mc.Infrastructure.ExporterAndImporter.Csv.csproj [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Magicodes.IE.Csv" Version="2.6.4" />
```
Comment (doddgu) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/Masa.Mc.Infrastructure.ExporterAndImporter.Csv.csproj:10]: Why using Magicodes.IE.Csv?

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/Utility/ExportHelper.cs [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+/// <summary>
+/// Csv导出辅助类
+/// </summary>
+/// <typeparam name="T"></typeparam>
+public class ExportHelper<T> where T : class
+{
+    private readonly Type _type;
+
+    /// <summary>
```
Comment (doddgu) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/Utility/ExportHelper.cs:14]: Remove the comment

File: src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/Utility/ImportHelper.cs [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+namespace Masa.Mc.Infrastructure.ExporterAndImporter.Csv.Utility
+{
+    /// <summary>
+    /// 导入辅助类
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class ImportHelper<T> : IDisposable where T : class, new()
+    {
+        /// <summary>
```
Comment (doddgu) [inline src/Infrastructure/Masa.Mc.Infrastructure.ExporterAndImporter.Csv/Utility/ImportHelper.cs:12]: remove the comment

File: src/Services/Masa.Mc.Service/Application/Channels/ChannelCommandHandler.cs [hunk 15]
```diff
@@ -45,8 +47,16 @@ public async Task DeleteAsync(DeleteChannelCommand createCommand)
             throw new UserFriendlyException("channel not found");
         if (await _messageTaskRepository.FindAsync(x => x.ChannelId == createCommand.ChannelId, false) != null)
         {
-            throw new UserFriendlyException("There are message tasks in use and cannot be deleted");
+            throw new UserFriendlyException("There are channel in use and cannot be deleted");
```
Comment (doddgu) [inline src/Services/Masa.Mc.Service/Application/Channels/ChannelCommandHandler.cs:50]: channel -> channels

File: src/Services/Masa.Mc.Service/Domain/Channels/Services/ChannelDomainService.cs [hunk 16]
```diff
@@ -44,7 +44,7 @@ protected virtual async Task ValidateChannelAsync(string code, Guid? expectedId
         var dict = await _repository.FindAsync(d => d.Code == code);
         if (dict != null && dict.Id != expectedId)
         {
-            throw new UserFriendlyException("Template code cannot be repeated");
+            throw new UserFriendlyException("channel code cannot be repeated");
```
Comment (doddgu) [inline src/Services/Masa.Mc.Service/Domain/Channels/Services/ChannelDomainService.cs:47]: channel -> Channel

File: src/Services/Masa.Mc.Service/Domain/MessageTemplates/EventHandler/SmsTemplateSynchroEventHandler.cs [hunk 17]
```diff
@@ -22,6 +22,7 @@ public class SmsTemplateSyncEventHandler
     }
 
     [EventHandler]
+
```
Comment (doddgu) [inline src/Services/Masa.Mc.Service/Domain/MessageTemplates/EventHandler/SmsTemplateSynchroEventHandler.cs:25]: remove the blank line.

**Review by doddgu — CHANGES_REQUESTED** [review]

**Review by doddgu — APPROVED** [review]

**Pull request merged**